### PR TITLE
[improve][broker][PIP-318] Support not retaining null-key message during topic compaction

### PR DESCRIPTION
**Pull request by coderzc**
PIP: #21541 

### Motivation

Implement #21541 

### Modifications

* Add config `topicCompactionRemainNullKey` in the broker.conf, the default value is false.
* If not retaining null-key message, add an empty single message placeholder to result when rebatching the message.
* If the message with null-key, then add a null-key value to the result list of extractIdsAndKeysAndSize and record the delete event.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java [hunk 1]
```diff
@@ -129,9 +139,14 @@ public static Optional<RawMessage> rebatchMessage(RawMessage msg,
...
                 if (!singleMessageMetadata.hasPartitionKey()) {
-                    messagesRetained++;
-                    Commands.serializeSingleMessageInBatchWithPayload(singleMessageMetadata,
-                                                                      singleMessagePayload, batchBuffer);
+                    if (retainNullKey) {
+                        messagesRetained++;
+                        Commands.serializeSingleMessageInBatchWithPayload(singleMessageMetadata,
+                                singleMessagePayload, batchBuffer);
+                    } else {
+                        Commands.serializeSingleMessageInBatchWithPayload(emptyMetadata,
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java:147]: Why do we write empty metadata here? Just to refill the batch? why we can't change the batch size?

File: pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java [hunk 2]
```diff
@@ -144,6 +146,12 @@ private void phaseOneLoop(RawReader reader,
                                     deleteCnt++;
                                     mxBean.addCompactionRemovedEvent(reader.getTopic());
                                 }
+                            } else {
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java:149]: Why the e is null? I didn't find we put a null to the list.

File: pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java [hunk 3]
```diff
@@ -75,25 +75,35 @@ public static List<ImmutableTriple<MessageId, String, Integer>> extractIdsAndKey
...
-                idsAndKeysAndSize.add(ImmutableTriple.of(id,
-                        smm.getPartitionKey(),
-                        smm.hasPayloadSize() ? smm.getPayloadSize() : 0));
+            if (!smm.isCompactedOut()) {
+                if (smm.hasPartitionKey()) {
+                    idsAndKeysAndSize.add(ImmutableTriple.of(id,
+                            smm.getPartitionKey(),
+                            smm.hasPayloadSize() ? smm.getPayloadSize() : 0));
+                } else {
+                    idsAndKeysAndSize.add(null);
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java:84]: please avoid add null to the list. :)

Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java:84]: +1
https://github.com/apache/pulsar/blob/2322004069f23ecca1a12a4ced898fed854275fb/pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java#L135-L147

File: pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java (hunk 1 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java:147]: The empty metadata set `compactedOut=true`, the client will filter out this message.

Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java:147]: > why we can't change the batch size?
We need to calculate the correct bantchIndex according to batch size and index of message.

File: pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java (hunk 3 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java]: Fixed.

File: pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java [hunk 4]
```diff
@@ -129,9 +135,14 @@ public static Optional<RawMessage> rebatchMessage(RawMessage msg,
                                                       msg.getMessageIdData().getPartition(),
                                                       i);
                 if (!singleMessageMetadata.hasPartitionKey()) {
-                    messagesRetained++;
-                    Commands.serializeSingleMessageInBatchWithPayload(singleMessageMetadata,
-                                                                      singleMessagePayload, batchBuffer);
+                    if (retainNullKey) {
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java:138]: how about move this logic to filter?

Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java:138]: If `topicCompactionRemainNullKey=true` then we also need to retain `<null, "">` message but filter will can't retain that message with an empty payload.

**Review by Technoboy- — APPROVED** [review]

**Review by mattisonchao — APPROVED** [review]

File: conf/broker.conf [hunk 5]
```diff
@@ -538,6 +538,9 @@ brokerServiceCompactionThresholdInBytes=0
 # If the execution time of the compaction phase one loop exceeds this time, the compaction will not proceed.
 brokerServiceCompactionPhaseOneLoopTimeInSeconds=30
 
+# Whether retain null-key message during topic compaction
+topicCompactionRemainNullKey=false
```
[the remark this answers is not in the document]
Comment (merlimat) [reply conf/broker.conf:542]: @coderzc I just saw this typo in the config key :)

[the remark this answers is not in the document]
Comment (coderzc) [reply conf/broker.conf:542]: Sorry, I open a new PR to fix it. https://github.com/apache/pulsar/pull/21690

**Pull request merged**